File: framework/languages/dotnet/e2e/AutomaticTurnDispatch/Server/Delay/DelayHostFactory.cs
using Microsoft.Extensions.Configuration;

using Systems.Zlink;
using AutomaticTurnDispatch.Shared;
using Zlink.Framework.AspNetCore;

namespace AutomaticTurnDispatch.Server.Delay;

internal static class DelayHostFactory
{
    public static WebApplication Create(string[] args)
    {
        var options = DelayOptions.Parse(args);
        Directory.CreateDirectory(options.LogDir);

        var builder = WebApplication.CreateBuilder(args);
        builder.Configuration.Sources.Clear();
        builder.Configuration.AddInMemoryCollection();
        builder.Logging.ClearProviders();
        builder.Logging.AddSimpleConsole(console =>
        {
            console.SingleLine = true;
            console.TimestampFormat = "HH:mm:ss.fff ";
        });
        builder.WebHost.UseUrls(options.HttpUrl);
        builder.Services.AddSingleton(new EvidenceStore(options.Rid, options.EvidenceFile));
        builder.Services.AddSingleton(new NodeOptions(options.Rid));
        builder.Services.AddZLinkFramework(framework =>
        {
            //  This E2E host is not started inside a memory-limited
            //  container. Supply a deterministic finite limit so the
            //  default Auto HWM contract does not depend on the host.
            framework.ConfigureInboundDispatch().ProcessMemoryLimitBytes =
                1UL * 1024 * 1024 * 1024;
            framework.AddHandlersFromAssemblyOf(typeof(Program));
            var mesh = framework.AddRouteMesh(AutomaticTurnDispatchNames.DelayChannel)
                .Listen(options.DelayEndpoint)
                .SetRoutingId(RoutingId.From(options.Rid));
            mesh.Channel(AutomaticTurnDispatchNames.DelayChannel).Server()
                .AddRequestHandler<DelayHandler, DelayReq, DelayRes>("DelayReq");
        });

        var app = builder.Build();
        app.MapGet("/health", () => Results.Ok(new { status = "ready", role = "delay", options.Rid }));
        app.MapGet("/evidence", (EvidenceStore evidence) => Results.Ok(evidence.Snapshot()));
        return app;
    }
}

File: framework/languages/node/e2e/AutomaticTurnDispatch/Server/Delay/delay-host-factory.ts
import { Module } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { ZLinkModule, zlinkFramework } from '@zlink-systems/nestjs';
import { AutomaticTurnDispatchNames } from '../../Shared/messages';
import { EvidenceStore } from './Support/evidence-store';
import { closeHttpServer, startHttpServer } from './Support/http-server';
import { createAutomaticTurnConfigurationModule } from '../../configuration';
import { DELAY_OPTIONS, validateDelayOptions } from './Configuration/delay-options';
import type { DelayOptions } from './Configuration/delay-options';
import { DelayHandler } from './Handlers/delay-handler';

export async function startDelayHost(): Promise<void> {
  let stopping = false;
  const configuration = createAutomaticTurnConfigurationModule(DELAY_OPTIONS, validateDelayOptions);

  class DelayModule {}
  Module({
    imports: [
      configuration,
      ZLinkModule.forRootFactory({
        imports: [configuration],
        inject: [DELAY_OPTIONS],
        useFactory: (value: unknown) => {
          const options = value as DelayOptions;
          const builder = zlinkFramework();
          const delay = builder.addRouteMesh(AutomaticTurnDispatchNames.delayChannel)
            .listen(options.delayEndpoint)
            .routingId(options.rid);
          delay.channel(AutomaticTurnDispatchNames.delayChannel).server()
            .addRequestHandler('DelayReq', DelayHandler);
          return builder.build();
        }
      })
    ],
    providers: [
      {
        provide: EvidenceStore,
        inject: [DELAY_OPTIONS],
        useFactory: (options: DelayOptions) => new EvidenceStore(options.rid, options.evidenceFile)
      },
      DelayHandler
    ]
  })(DelayModule);

  const app = await NestFactory.createApplicationContext(DelayModule, { logger: false });
  const options = app.get<DelayOptions>(DELAY_OPTIONS);
  const evidence = app.get(EvidenceStore);
  const server = await startHttpServer(options.httpUrl, [
    { method: 'GET', path: '/health', handle: () => ({ status: 'ready', role: 'delay', rid: options.rid }) },
    { method: 'GET', path: '/evidence', handle: () => evidence.snapshot() },
    {
      method: 'POST',
      path: '/shutdown',
      handle: () => {
        stopping = true;
        return { status: 'stopping' };
      }
    }
  ]);
  while (!stopping) {
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  await closeHttpServer(server);
  await app.close();
}
